Fixed Hard Coupling (com_users | plg_users_joomla)

carlitorweb opened the pull request:
Pull Request for Issue #20249

### Summary of Changes
Added `onContentPrepareForm` event to the plugin for doing the form require alterations to the passwords fields.

### Testing Instructions
Go to the administration and create a new user. Check the form that is shown need have the passwords fields with the attribute "require" if the plugin **User - Joomla!** have the Notification Mail to User option as No.

### Expected result
Passwords field with the "require" attribute 

### Actual result
Passwords field with the "require" attribute 

### Documentation Changes Required
No

carlitorweb committed:
Fixed Hard Coupling (com_users | plg_users_joomla)

zero-24 commented on plugins/user/joomla/joomla.php:
```diff
@@ -34,6 +34,40 @@ class PlgUserJoomla extends JPlugin
+	 * @param   mixed  $data  The associated data for the form.
+	 *
+	 * @return  boolean
+	 *
```
please add `@since  __DEPLOY_VERSION__` ;)

carlitorweb replied:
always forget something :D

zero-24 commented on plugins/user/joomla/joomla.php:
```diff
@@ -34,6 +34,40 @@ class PlgUserJoomla extends JPlugin
+	 */
+	public function onContentPrepareForm($form, $data)
+	{
+		if (!($form instanceof JForm))
```
hmm IMO this check can be dropped as if we get here without a valid form something else is broken anyway. I have not checked it but IIRC `$this->_subject->setError('JERROR_NOT_A_FORM');` would not work anyway.

carlitorweb replied:
> if we get here without a valid form something else is broken anyway

I thought same, but I saw we use it inside another plugin, and no wanted to drop something without knowing the "exactly why"

zero-24 commented on plugins/user/joomla/joomla.php:
```diff
@@ -34,6 +34,40 @@ class PlgUserJoomla extends JPlugin
+	 * @param   JForm  $form  The form to be altered.
+	 * @param   mixed  $data  The associated data for the form.
+	 *
+	 * @return  boolean
```
It seams drone don't likes this line ;)

```
FILE: /drone/src/github.com/joomla/joomla-cms/plugins/user/joomla/joomla.php
--------------------------------------------------------------------------------
FOUND 1 ERROR(S) AFFECTING 1 LINE(S)
--------------------------------------------------------------------------------
 43 | ERROR | Return comment requires a blank newline after it
```
http://ci.joomla.org/joomla/joomla-cms/5010/4

carlitorweb committed:
Code review. Dropped instanceof JForm check

carlitorweb commented:
Done @zero-24 and thank you

zero-24 commented on plugins/user/joomla/joomla.php:
```diff
@@ -34,6 +34,34 @@ class PlgUserJoomla extends JPlugin
+	 *
+	 * @return  boolean
+	 *
+	 * @since  3.8
```
Please use `__DEPLOY__VERSION__` this way the release script replace it with the correct version where that method get committed in :)

carlitorweb committed:
Code review

Quy commented on plugins/user/joomla/joomla.php:
```diff
@@ -34,6 +34,34 @@ class PlgUserJoomla extends JPlugin
+	 *
+	 * @return  boolean
+	 *
+	 * @since  __DEPLOY__VERSION__
```
Align. Add one more space.

carlitorweb committed:
Code review

zero-24 commented:
I have just tested this and found two issue. In case there is a validation error $data is no stdClass but just an empty array. And the second is what happen when the plugin is disabled. Currently the password field is not required. But it needs to be required than.

## How to reproduce (First Issue)

- set error reporting to max
- create a user with the username `test`
- try to create another user with the same username
- error pops up.
- when you debug you notice that the $data array is empty.

## How to reproduce (Second Issue)

- Disable the plugin
- try to create a user
- pw field is not required.

## How to solve the issues

- change default in com_users XML file that the pw's are always required
- restructure the plugin check so it disable the required status if running.

carlitorweb commented:
>Disable the plugin

@zero-24 if you disabled the plugin, others things like the login logic will break, among other things. In fact, there is a string warning about this. In case this plugin is disabled because another plugin will handle User synchronization, then we no need more the "Notification Mail to User", and we no need check for this.

> try to create another user with the same username

We can add an `is_object()` check inside the condition for avoiding this

carlitorweb committed:
Fixed issue when username is same

Try to create another user with the same username. Error pops up. When you debug you notice that the $data array is empty

carlitorweb commented:
I do not understand why the drone fail here?

carlitorweb committed:
Fix drone

carlitorweb committed:
Fix drone

zero-24 commented:
> In case this plugin is disabled because another plugin will handle User synchronization, then we no need more the "Notification Mail to User", and we no need check for this.

Well my point is that the password fields are than not required in that case ;)
So my suggestion was to change the logic so the fields are required allways but not required with the setting is enabeld ;)

But I'm also fine merge this without that change and when this got accepted i do a PR that would handle the proposed solution.

carlitorweb commented:
@Quy can you give me a hand here? This code is marked as a conflict for the merge, was moved to the plugin for solved the issue. Or maybe I understand wrong this.

Quy commented:
I think you have to update your branch since there were changes to `administrator/components/com_users/models/user.php` 4 days ago.

carlitorweb committed:
Merge branch 'staging' into hard_coupling_users

carlitorweb commented:
Thank you @Quy

carlitorweb committed:
CS fixed

SharkyKZ commented on plugins/user/joomla/joomla.php:
```diff
@@ -34,6 +34,34 @@ class PlgUserJoomla extends JPlugin
+	 *
+	 * @return  boolean
+	 *
+	 * @since   __DEPLOY__VERSION__
```
It's `__DEPLOY_VERSION__` (one underscore between words).

SharkyKZ commented on plugins/user/joomla/joomla.php:
```diff
@@ -34,6 +34,34 @@ class PlgUserJoomla extends JPlugin
+		if ($name === 'com_users.user')
+		{
+			// Passwords fields are required when mail to user is set to No
+			if (is_object($data) && $data->id === 0 && $this->params->get('mail_to_user', '1') === '0')
```
Change `mail_to_user` param fallback value and comparison value to be integer and comparison to be not type-safe:
`$this->params->get('mail_to_user', 1) == 0` or `!$this->params->get('mail_to_user', 1)`.

carlitorweb commented:
Thank you @SharkyKZ

carlitorweb committed:
Codereview

Quy commented:
Create a duplicate user.
An error message is displayed.
`Password` and `Confirm Password` are no longer marked required.

#14851 and #20313 could be related to this issue.

SharkyKZ commented:
Data from user state is an array https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_users/models/user.php#L172.

This seems to be the case everywhere else.

carlitorweb committed:
Change password field required as default

carlitorweb committed:
Restructure the plugin check so it disable the required status if running

carlitorweb committed:
Pasword field is not required when we edit a user

carlitorweb commented:
I made what @SharkyKZ suggest before:

> change default in com_users XML file that the pw's are always required
restructure the plugin check so it disable the required status if running.

Please, can you test again @Quy . Also, as the password field is required by default, we need to remove this, when we edit a user. 

PD: Sorry for the delay

Quy commented:
On the backend, `Password` and `Confirm Password` are not required (right). After an error, these fields are marked required (wrong).

carlitorweb commented:
>On the backend, `Password` and `Confirm Password` are not required (right)

So, change the pw´s fields required by default is wrong?

Quy commented:
On the backend, leave the password blank for the system to auto generate the password. Without PR, the 2 password fields are not required. With PR, the password fields are not required initially, but after a duplicate error, they are marked required.

carlitorweb committed:
Revert "Pasword field is not required when we edit a user"

This reverts commit 8cad84d628967af474e870746253af640dca865f.

carlitorweb committed:
Revert "Restructure the plugin check so it disable the required status if running"

This reverts commit 5f504ccf2604a92d26ab00df8c669995ac8b77e8.

carlitorweb committed:
Revert "Change password field required as default"

This reverts commit 86ee607773fe743df586e56974adc444255e16c8.

carlitorweb commented:
@Quy now I got it. Sorry for be slow, sometime my English is not so good as I want it. 
So, with this PR and plugin User - Joomla, with the `Notification Mail to User` set on `No` passwords fields are required (Correct operation). But if you save a duplicate user, then passwords fields after get the error, are not required (Wrong operation, need stay as required).

Checking this, somehow $options['load_data'] come with the value false when you save a duplicate user, https://github.com/joomla/joomla-cms/blob/58cd2dcb3aebdb17752cf71c169eebf8c1f59e11/libraries/src/MVC/Model/FormModel.php#L230 then, preprocessForm(...) trigger the plugin passing an empty $data, so that make the check inside onContentPrepareForm event fail.

Quy commented:
Youre English fine to me. 👍 

Please read the following about password on the backend: https://github.com/joomla/joomla-cms/pull/20247#issuecomment-385167234

I was wrong with my report. I think your PR is fine, but I will have to test again to be sure.

carlitorweb commented:
> Please read the following about password on the backend

Thank you

SharkyKZ commented:
The issue is that on validation error returned data is an array. This could probably be fixed in core, but for now you can remove `is_object()` check and cast data as object.

carlitorweb committed:
Casting $data

Quy commented:
Tested @ https://github.com/joomla/joomla-cms/pull/20275/commits/9e333a998656c0e97b979d43cf66d0d60cc501b1

> PHP Notice:  Undefined property: stdClass::$id in \plugins\user\joomla\joomla.php on line 58

SharkyKZ commented:
@Quy how do you get this notice?

Quy commented:
Check in the PHP error log file.

SharkyKZ commented:
FormController never loads form data so whatever form data check is performed in plugin is not valid when saving.
[https://github.com/joomla/joomla-cms/blob/58cd2dcb3aebdb17752cf71c169eebf8c1f59e11/libraries/src/MVC/Controller/FormController.php#L688](https://github.com/joomla/joomla-cms/blob/58cd2dcb3aebdb17752cf71c169eebf8c1f59e11/libraries/src/MVC/Controller/FormController.php#L688)
Assuming this is intended and no changes should be made to models or controllers, see `System - Fields` plugin for a solution:
[https://github.com/joomla/joomla-cms/blob/58cd2dcb3aebdb17752cf71c169eebf8c1f59e11/plugins/system/fields/fields.php#L274-L287](https://github.com/joomla/joomla-cms/blob/58cd2dcb3aebdb17752cf71c169eebf8c1f59e11/plugins/system/fields/fields.php#L274-L287)

carlitorweb commented:
Thank you @SharkyKZ and @Quy  .

carlitorweb committed:
Casting $data

Quy commented:
Sorry I gave you wrong information prior. Here is the right information:

On the back end, when `Notification Mail to User` is `No`, then `Password` and `Confirm Password` should be marked with `*` and be required. 

Your PR, however, after an error, these 2 fields are no longer marked with `*`.

SharkyKZ commented:
This isn't right yet. On form display, `$data` is always populated. It's normally an object but an array after returning from error. Data from input (`$jformData`) is not available on form display.

On save, `$data` is an empty array because `FormController` does not load data. So on save you need to use data from input (`$jformData`).

This is indeed #14851.

The same issue exists without PR. When `Notification Mail to User`  is set to `No` password fields appear to be required on form display but aren't validated on save. To test this, remove `required` attribute and class from markup using browser's developer tools and submit the form without passwords. The user will be saved.

carlitorweb commented:
The real fix is try to handle this from the model or the controller. But I can bet will be B/C issue.

> Your PR, however, after an error, these 2 fields are no longer marked with *

Is true, I do not see this. However, if you try to save again, you get a warning telling you, the password field is required.

carlitorweb commented:
@Quy I does not understand why, when we get the error the `*` gone. Even if you save, you got the warning saying the password is required, but no put the field as required in the HTML markup. Can you give me a hand with this? I try to find where I miss something, but no luck for now

SharkyKZ commented:
It's because of the check in the plugin. Replace this:
```
if (is_array($data) && $jformData)
{
	$data = (object) $jformData;
}
```
With this:
```
if ($jformData && !$data) 
 { 
 	$data = $jformData; 
 } 
  
if (is_array($data)) 
{ 
 	$data = (object) $data; 
} 
```

carlitorweb committed:
Cast $data

carlitorweb commented:
Thank you @SharkyKZ , that was But or I need close all here and go to watch a movie, or is almost the same I had in the check before :D . Anyway, thank you for the help

Quy commented on plugins/user/joomla/joomla.php:
```diff
@@ -34,6 +34,49 @@ class PlgUserJoomla extends JPlugin
+
+		if ($name === 'com_users.user')
+		{
+			// In case there is a validation error (like duplicated user), $data is no a stdClass but just an empty array
```
Change `no` to `not`

Quy commented on plugins/user/joomla/joomla.php:
```diff
@@ -34,6 +34,49 @@ class PlgUserJoomla extends JPlugin
+			}
+
+			// Passwords fields are required when mail to user is set to No
+			if (isset($data->id) && (int) $data->id == 0 && !$this->params->get('mail_to_user', 1))
```
Can be simplified as follows: `if (empty($data->id) && !$this->params->get('mail_to_user', 1))`

carlitorweb committed:
Code review

carlitorweb commented:
Done, thank @Quy

ggppdk commented on plugins/user/joomla/joomla.php:
```diff
@@ -34,6 +34,49 @@ class PlgUserJoomla extends JPlugin
+			if ($jformData && !$data)
+			{
+				$data = $jformData;
+			}
```
Can be written like:
```php
	if (!$data)
	{
		$data = JFactory::getApplication()->input->get('jform', array(), 'array');
	}
```

carlitorweb replied:
look better, thank you

mbabker commented on plugins/user/joomla/joomla.php:
```diff
@@ -34,6 +34,49 @@ class PlgUserJoomla extends JPlugin
+		if ($name === 'com_users.user')
+		{
+			// In case there is a validation error (like duplicated user), $data is not a stdClass but just an empty array
+			// TODO: Check inside the FormModel line 230, for put the right associated data for the form
```
Really shouldn't make a reference to a specific line number in a code comment.  If the code's updated, this reference goes away.

carlitorweb replied:
Is true, thank you @mbabker my mistake, sorry

SharkyKZ replied:
The first comment should also be updated. `$data` is an empty array on save. After returning from error, `$data` is an array but populated.

@mbabker can you comment on form data issue? `FormController` does not load form data by default, making it unavailable in `onContentPrepareForm` when saving. Code based on `Form::getValue()` also fails. Is this correct behavior? Should we use data from input instead?

carlitorweb committed:
Code review

carlitorweb committed:
Code review

Quy commented:
Related issue #17700 about empty $data.

carlitorweb commented:
This is the correct way then to do this or need a different way?

SharkyKZ commented:
The code works fine now.

Still thinking about @zero-24's suggestion to make password fields required by default. It would make sense not to auto-generate passwords when plugin is disabled, when mail to user option in plugin is disabled or when sending passwords is disabled in com_users.

carlitorweb commented:
Ok, let do the suggestion of @zero-24

brianteeman commented:
It never makes sense in the admin to make creating a password required.

mbabker commented:
> It never makes sense in the admin to make creating a password required.

For creating a new account, I'd say it would make sense to default it to required and conditionally unrequire it based on the other settings.  Otherwise you're left in a spot where admin creates an account without setting a password and the new user's only way of getting into the account is through the "forgot my password" interface.  I don't think that's necessarily the best new user experience, but we all know what opinions are like.

carlitorweb commented:
This is a valid argument imo [#20247 (comment)](https://github.com/joomla/joomla-cms/pull/20247#issuecomment-385167234)

franz-wohlkoenig commented:
@carlitorweb are the Test Instructions up-to-date?

carlitorweb commented:
@Fran so far, yes

franz-wohlkoenig commented:
"Notification Mail to User"-option is "No", both Password-Fields have a "*". I'm confused about the Test Instructions. 

Also "Expected" and "Actual Result" have the same Description.

SharkyKZ commented:
@franz-wohlkoenig Behavior before and after patch should be the same.

SharkyKZ commented:
I have tested this item :white_check_mark: successfully on fe03395bce8b16d3db6685e938b7eb4fc7f228c6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20275">issues.joomla.org/tracker/joomla-cms/20275</a>.</sub>

Commit:
Merge branch 'staging' into hard_coupling_users

Commit:
Merge branch 'staging' into hard_coupling_users

Quy commented:
I have tested this item :white_check_mark: successfully on b633bbfbdc99d9a8cf843ca6cae0e06c4c68d855<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20275">issues.joomla.org/tracker/joomla-cms/20275</a>.</sub>

Quy commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20275">issues.joomla.org/tracker/joomla-cms/20275</a>.</sub>

HLeithner committed:
Merge branch 'staging' into hard_coupling_users

HLeithner commented:
Thanks
